### PR TITLE
[receiver/mongodbatlas] Add Events API Support

### LINKED ISSUE #18061: [receiver/mongodbatlas] Add Events API Support
### Component(s)

receiver/mongodbatlas

### Is your feature request related to a problem? Please describe.

Looking to add to the receiver to start collecting events from [this API](https://www.mongodb.com/docs/atlas/reference/api/events-projects-get-all/). This endpoint contains a multitude of types of information about `Host Restarts`, `Replica set elected a new primary`, `Alert configuration disabled `, ` Network: Bytes In has gone above 5000 bytes (avg/sec) `, `User joined the project`, etc. 

I feel like this type of data would be useful for end users to use this data to track rapid restarts and potentially create alerts. 

### Describe the solution you'd like

I think a good solution for this is to add another section of configuration along with `alerts`, `logs`, and `metrics` which could be called `events`.  

I also think it nice to include a project filtering mechanism in the config along with an event type filtering on the type of event, by default all should be collected. 

Example Configuration:

```yaml
receivers:
  mongodbatlas:
    public_key: <redacted>
    private_key: <redacted>
    events:
      enabled: true
      # only collect AUTH_LOGIN_FAIL and HOST_RESTARTED events
      types: [AUTH_LOGIN_FAIL, HOST_RESTARTED]
      projects:
      - name: Project 0
        include_clusters: [Cluster0]
      poll_interval: 1m
    # use of a storage extension is recommended to reduce chance of duplicated events
    storage: file_storage
```

### Describe alternatives you've considered

I thought about including this in the `alerts` portion but ultimately decided against it because not all these events are "alerts" but more info level information such as user additions, host restarts, etc.

### Additional context

I'm happy to start working on this and willing to be assigned if the feature is acceptable. Happy to hear further suggestions about configuration/overall usability.

### PR DESCRIPTION
**Pull request by schmikei**
**Description:** Added support for collecting Events from the MongoDB Atlas API

Sample Configuration

```yaml
mongodbatlas:
  public_key: "my-public-key"
  private_key: "my-private-key"
  events:
   projects:
      - name: Project 0
    poll_interval: 1m
    types: [ALERT_CONFIG_CHANGED_AUDIT]
    page_size: 100
    max_pages: 25
```

**Link to tracking Issue:** Resolves #18061 

**Testing:** Unit Testing

**Documentation:** Added more information on configuration to the README

### CODE REVIEW

**Comment by schmikei** [comment]
@djaglowski here's my first stab at implementing this based off your guys' feedback in #18237, happy to iterate on the solution as well!

File: receiver/mongodbatlasreceiver/README.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ MongoDB Atlas [Documentation](https://www.mongodb.com/docs/atlas/reference/api/l
 - `public_key` (required for metrics, logs, or alerts in `poll` mode)
 - `private_key` (required for metrics, logs, or alerts in `poll` mode)
 - `granularity` (default `PT1M` - See [MongoDB Atlas Documentation](https://docs.atlas.mongodb.com/reference/api/process-measurements/))
-- `storage` configure the component ID of a storage extension. If specified, alerts `poll` mode will utilize the extension to ensure alerts are not duplicated after a collector restart.
+- `storage` configure the component ID of a storage extension. If specified, both alerts `poll` mode and `events` collection will utilize the extension to ensure alerts are not duplicated after a collector restart.
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/README.md:27]: Let's reword this to untangle alerts and events. Maybe:
```suggestion
- `storage` (optional) The component ID of a storage extension which can be used when polling for `alerts` or `events` . The storage extension prevents duplication of data after a collector restart by remembering which data were previously collected.
```

File: receiver/mongodbatlasreceiver/README.md [hunk 2]
```diff
@@ -55,6 +55,17 @@ MongoDB Atlas [Documentation](https://www.mongodb.com/docs/atlas/reference/api/l
...
+- `events`
+  - `projects`
+    - `name` Name of the Project to discover events from
+  - `poll_interval` (default `1m`)
+    - How often the receiver will poll the Events API for new events.
+  - `page_size` (default `100`)
+    - This is the number of events that will be processed per request to the MongoDB Atlas API.
+  - `max_pages` (default `25`)
+    - This will limit how many pages of events the receiver will request from the MongoDB Atlas API for each project.
+  - `types` (default [])
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/README.md:67]:
```suggestion
  - `types` (defaults to all types of events)
```
or something to that effect

File: receiver/mongodbatlasreceiver/README.md [hunk 3]
```diff
@@ -55,6 +55,17 @@ MongoDB Atlas [Documentation](https://www.mongodb.com/docs/atlas/reference/api/l
...
+    - This is a list of [event types](https://www.mongodb.com/docs/atlas/reference/api/events-orgs-get-all/#event-type-values) that the receiver will request from the API. If not specified then the receiver will collect all types of events.
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/README.md:68]:
```suggestion
    - This is a list of [event types](https://www.mongodb.com/docs/atlas/reference/api/events-orgs-get-all/#event-type-values) that the receiver will request from the API. If specified, the receiver will collect only the indicated types of events.
```

File: receiver/mongodbatlasreceiver/config.go [hunk 4]
```diff
@@ -39,6 +39,7 @@ type Config struct {
 	Granularity                             string                       `mapstructure:"granularity"`
 	Metrics                                 metadata.MetricsSettings     `mapstructure:"metrics"`
 	Alerts                                  AlertConfig                  `mapstructure:"alerts"`
+	Events                                  *EventsConfig                `mapstructure:"events,omitempty"`
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/config.go:42]: Out of curiosity, why `omitempty` for this one when others do not have the tag?

File: receiver/mongodbatlasreceiver/config.go [hunk 5]
```diff
@@ -109,6 +119,9 @@ func (c *Config) Validate() error {
 
 	errs = multierr.Append(errs, c.Alerts.validate())
 	errs = multierr.Append(errs, c.Logs.validate())
+	if c.Events != nil {
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/config.go:122]: Why do we need to check if this one is nil, but not Alerts and Logs?

File: receiver/mongodbatlasreceiver/config.go [hunk 6]
```diff
@@ -194,3 +207,12 @@ func (a AlertConfig) validateListenConfig() error {
...
 	return errs
 }
+
+func (e EventsConfig) validate() error {
+	var errs error
+	if len(e.Projects) == 0 {
+		errs = multierr.Append(errs, errNoProjects)
+	}
+
+	return errs
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/config.go:217]:
```suggestion
	if len(e.Projects) == 0 {
		return errNoProjects
	}
	return nil
```

Why use multierr here? There is only one possible error.

File: receiver/mongodbatlasreceiver/events.go [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+	return er.checkpoint(ctx)
+}
+
+func (er *eventsReceiver) startPolling(ctx context.Context, host component.Host) error {
+	storageClient, err := adapter.GetStorageClient(ctx, host, er.storageID, er.id)
+	if err != nil {
+		return fmt.Errorf("failed to set up storage: %w", err)
+	}
+	er.storageClient = storageClient
+	err = er.loadCheckpoint(ctx)
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:120]: This isn't really polling related. Let's pull it out into the `Start` method, prior to calling `startPolling`.

File: receiver/mongodbatlasreceiver/events.go [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+		parseOptionalAttributes(&attrs, event)
+	}
+
+	return logs
+}
+
+func (er *eventsReceiver) checkpoint(ctx context.Context) error {
+	if er.storageClient == nil {
+		return errors.New("missing non-nil storage client")
+	}
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:239]: Not sure this can happen with current code. If storage is unspecified we will get a dummy client and shouldn't need to worry about this.

File: receiver/mongodbatlasreceiver/events.go [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+	if err != nil {
+		return fmt.Errorf("unable to write checkpoint: %w", err)
+	}
+	return er.storageClient.Set(ctx, eventStorageKey, marshalBytes)
+}
+
+func (er *eventsReceiver) loadCheckpoint(ctx context.Context) error {
+	if er.storageClient == nil {
+		return nil
+	}
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:250]: Same

File: receiver/mongodbatlasreceiver/events.go [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+	cBytes, err := er.storageClient.Get(ctx, eventStorageKey)
+	if err != nil || cBytes == nil {
+		er.record = &eventRecord{}
+		return nil
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:254]: Should we log the error, since we aren't returning it. (I agree with not returning it, since we can proceed without the checkpoint)

File: receiver/mongodbatlasreceiver/events.go [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+	storageID     *component.ID // ID of the storage extension component
+	storageClient storage.Client
+	cfg           *Config
+	consumer      consumer.Logs
+
+	maxPages     int
+	pageSize     int
+	pollInterval time.Duration
+	wg           *sync.WaitGroup
+	record       *eventRecord
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:63]: Is this used for checkpointing? Worth a comment to describe how, if so.

File: receiver/mongodbatlasreceiver/events.go [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+	for _, event := range events {
+		resourceLogs := logs.ResourceLogs().AppendEmpty()
+		ra := resourceLogs.Resource().Attributes()
+		ra.PutStr("mongodbatlas.project.name", p.Name)
+		ra.PutStr("mongodbatlas.org.id", p.OrgID)
+		ra.PutStr("mongodbatlas.group.id", event.GroupID)
+
+		if event.ReplicaSetName != "" {
+			ra.PutStr("mongodbatlas.replica_set.name", event.ReplicaSetName)
+		}
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:205]: Should we separate events by group id and replicate set name, so we can put multiple logs on a resource, rather than emitting a resource for every log?

File: receiver/mongodbatlasreceiver/events.go [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+			er.logger.Error("unable to unmarshal event into body string", zap.Error(err))
+			continue
+		}
+		logRecord.Body().SetStr(string(bodyBytes))
+
+		// ISO-8601 formatted
+		ts, err := time.Parse(time.RFC3339, event.Created)
+		if err != nil {
+			er.logger.Warn("unable to interpret when an event was created, expecting a RFC3339 timestamp", zap.String("timestamp", event.Created), zap.String("event", event.ID))
+			continue
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:219]: Should we emit the event anyways with a "now" timestamp?

File: receiver/mongodbatlasreceiver/events.go [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+	return nil
+}
+
+func parseOptionalAttributes(m *pcommon.Map, event *mongodbatlas.Event) {
+	if event.AlertID != "" {
+		m.PutStr("alert.id", event.AlertID)
+	}
+
+	if event.AlertConfigID != "" {
+		m.PutStr("mongodbatlas.alert.config.id", event.AlertConfigID)
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:271]: Why have a prefix only for this attribute?

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go [hunk 15]
```diff
@@ -648,6 +648,45 @@ func (s *MongoDBAtlasClient) GetAlerts(ctx context.Context, groupID string, opts
...
+	// to grab from the API
+	EventTypes []string
+	// The oldest date to look back for the events
+	MinDate time.Time
+	// the newest time to accept events
+	MaxDate time.Time
+}
+
+// GetEvents returns the events specified for the set projects
+func (s *MongoDBAtlasClient) GetEvents(ctx context.Context, groupID string, opts *GetEventsOptions) (ret []*mongodbatlas.Event, nextPage bool, err error) {
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:667]: In the calling function, this is project ID, and group ID appears to mean something else.

Comment (schmikei) [reply receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:667]: Was just following the convention of the previous function signatures. The docs refer to these as group ids as well which is why I think that was initially chosen.

https://www.mongodb.com/docs/atlas/reference/api/events-projects-get-all/#request-path-parameters

Would you prefer the param be renamed to `projectID`?

File: receiver/mongodbatlasreceiver/events.go (hunk 12 above)
Comment (schmikei) [reply receiver/mongodbatlasreceiver/events.go:205]: Ahh I think that's a good idea 👍

File: receiver/mongodbatlasreceiver/config.go (hunk 4 above)
Comment (schmikei) [reply receiver/mongodbatlasreceiver/config.go:42]: Yeah I was thinking if there was some marshal/unmarshal chain of the config that this would be an issue. But I don't think it necessary after playing around a tad bit more in a couple test cases. Will remove the tag 👍

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go (hunk 15 above)
Comment (djaglowski) [reply receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:667]: Ok, that's fine. Just wanted to make sure it wasn't a mistake.

File: receiver/mongodbatlasreceiver/events.go [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+func (er *eventsReceiver) Start(ctx context.Context, host component.Host) error {
+	er.logger.Debug("Starting up events receiver")
+	storageClient, err := adapter.GetStorageClient(ctx, host, er.storageID, er.id)
+	if err != nil {
+		return fmt.Errorf("failed to set up storage: %w", err)
+	}
+	er.storageClient = storageClient
+	err = er.loadCheckpoint(ctx)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (schmikei) [reply receiver/mongodbatlasreceiver/events.go]: Changed it to log and not fail to start 👍

**Review by djaglowski — COMMENTED** [review]
Forgot to submit a couple comments. Hopefully easy ones

File: receiver/mongodbatlasreceiver/events.go [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+	maxPages     int
+	pageSize     int
+	pollInterval time.Duration
+	wg           *sync.WaitGroup
+	record       *eventRecord // this record is used for checkpointing last processed events
+	doneChan     chan bool
+}
+
+type eventRecord struct {
+	sync.Mutex
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:67]: You don't appear to be using this anywhere.

File: receiver/mongodbatlasreceiver/events.go [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+	if err != nil {
+		return fmt.Errorf("unable to write checkpoint: %w", err)
+	}
+	return er.storageClient.Set(ctx, eventStorageKey, marshalBytes)
+}
+
+func (er *eventsReceiver) loadCheckpoint(ctx context.Context) {
+	cBytes, err := er.storageClient.Get(ctx, eventStorageKey)
+
+	if cBytes == nil {
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:243]: Do we need to check `err` here?

File: receiver/mongodbatlasreceiver/events.go [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+		er.record = &eventRecord{}
+		return
+	}
+
+	if err != nil {
+		er.logger.Info("unable to load checkpoint from storage client, continuing without a previous checkpoint", zap.Error(err))
+		er.record = &eventRecord{}
+		return
+	}
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:252]: Typically check the error before any other return value, so move this block above the previous one.

File: receiver/mongodbatlasreceiver/events.go (hunk 17 above)
Comment (djaglowski) [reply receiver/mongodbatlasreceiver/events.go:67]: I'm not sure you actually need the mutex. Each instance of the receiver will have its own instance of `NextStartTime`. Within each instance of the receiver, you only access `NextStartTime` synchronously within each `poll`, and each `poll` blocks the next.

Comment (schmikei) [reply receiver/mongodbatlasreceiver/events.go:67]: Ahh you're right I'll remove it 👍 was concerned about the goroutine nature of it but you're right it doesn't need it

File: receiver/mongodbatlasreceiver/events.go [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+
+	maxPages     int
+	pageSize     int
+	pollInterval time.Duration
+	wg           *sync.WaitGroup
+	record       *eventRecord // this record is used for checkpointing last processed events
+	cancel       context.CancelFunc
+}
+
+type eventRecord struct {
```
Comment (djaglowski) [inline receiver/mongodbatlasreceiver/events.go:66]: Do we need this struct? (As opposed to just the time?)

Comment (schmikei) [reply receiver/mongodbatlasreceiver/events.go:66]: Primarily using it for the JSON marshal tag, I think it's required for that reason

**Review by djaglowski — APPROVED** [review]
